public/tracy/Tracy.hpp

```diff
@@ -149,10 +149,34 @@
 #define ZoneTransientN( varname, name, active ) tracy::ScopedZone varname( TracyLine, TracyFile, strlen( TracyFile ), TracyFunction, strlen( TracyFunction ), name, strlen( name ), TRACY_CALLSTACK, active )
 #define ZoneTransientNC( varname, name, color, active ) tracy::ScopedZone varname( TracyLine, TracyFile, strlen( TracyFile ), TracyFunction, strlen( TracyFunction ), name, strlen( name ), color, TRACY_CALLSTACK, active )
 
-#define ZoneScoped ZoneNamed( ___tracy_scoped_zone, true )
-#define ZoneScopedN( name ) ZoneNamedN( ___tracy_scoped_zone, name, true )
-#define ZoneScopedC( color ) ZoneNamedC( ___tracy_scoped_zone, color, true )
-#define ZoneScopedNC( name, color ) ZoneNamedNC( ___tracy_scoped_zone, name, color, true )
+#if defined(TRACY_ALLOW_SHADOW_WARNING)
+    #define SuppressVarShadowWarning(Expr) Expr
+#elif defined(__clang__)
+    #define SuppressVarShadowWarning(Expr) \
+        _Pragma("clang diagnostic push"); \
+        _Pragma("clang diagnostic ignored \"-Wshadow\""); \
+        Expr; \
+        _Pragma("clang diagnostic pop");
+#elif defined(__GNU__)
+    #define SuppressVarShadowWarning(Expr) \
+        _Pragma("GCC diagnostic push"); \
+        _Pragma("GCC diagnostic ignored \"-Wshadow\""); \
+        Expr; \
+        _Pragma("GCC diagnostic pop");
+#elif defined(_MSC_VER) 
+    #define SuppressVarShadowWarning(Expr) \
+        _Pragma("warning(push)"); \
+        _Pragma("warning(disable : 4456)"); \
+        Expr; \
+        _Pragma("warning(pop)");
```

Comment:
Since we have https://github.com/wolfpld/tracy/pull/1157, we'd better no end the macro with a semicolumn

Reply:
I see the point, at the same time I'm not sure if removing these semicolons here will do what we want. Since

A) We need a semicolon after `Expr` to properly separate it from `_Pragma`.
B) `_Pragma` doesn't need a semicolon after it to work correctly IIRC
C) This is not a macro intended for user use.

Maybe we should mark the macro as private, something like `__SuppressVarShadowWarning`?

Reply:
Right, but it will trigger `-Wextra-semi-stmt` on clang though. Introducing a new warning by suppressing another is a bit weird.

The only way I can think of on top of my head is by adding a `do {} while(0)` (test here https://godbolt.org/z/7b4hjv7Go ), but I don't know if this is something we want?

Reply:
> Yeah, that's why I was advocating for an opt-in toggle that would suppress these warnings right at the top of `Tracy.hpp` (and leave it enabled for the entire translation unit).

I think at that point it'd be better to let the users wrap it through their own headers :)